roles/uperf/tasks/setup.yml

```diff
@@ -0,0 +1,167 @@
+  register: resource_state
+
+- operator_sdk.util.k8s_status:
+    api_version: ripsaw.cloudbulldozer.io/v1alpha1
+    kind: Benchmark
+    name: "{{ meta.name }}"
+    namespace: "{{ operator_namespace }}"
+    status:
+      state: Building
+      complete: false
+  when: resource_state.resources[0].status.state is not defined
+
+- name: Get current state - If it has changed
+  k8s_facts:
+    api_version: ripsaw.cloudbulldozer.io/v1alpha1
+    kind: Benchmark
+    name: '{{ meta.name }}'
+    namespace: '{{ operator_namespace }}'
+  register: resource_state
+
+- name: Capture operator information
+  k8s_facts:
+    kind: Pod
+    api_version: v1
+    namespace: '{{ operator_namespace }}'
+    label_selectors:
+      - name = benchmark-operator
+  register: bo
+  
+- block:
+    #
+    # This block is for scale mode where client and server pods are spreaded
+    # across all eligible nodes
+    #
+  - name: List Nodes Labeled as Workers
+    k8s_info:
+      api_version: v1
+      kind: Node
+      label_selectors:
+        - "node-role.kubernetes.io/worker="
+    register: node_list
+    no_log: True
+  
+  - name: Isolate Worker Role Hostnames
+    set_fact:
+      worker_node_list: "{{ node_list | json_query('resources[].metadata.name') | list }}"
+  
+  - name: List Nodes Labeled with {{ workload_args.exclude_label }}
+    k8s_info:
+      api_version: v1
+      kind: Node
+      label_selectors:
+        - '{{ item }}'
+    with_items: "{{ workload_args.exclude_labels }}"
+    register: exclude_node_list
+    when: workload_args.exclude_labels is defined and workload_args.exclude_labels | length > 0
+  
+  - name: Isolate Worker Role Hostnames for label  {{ workload_args.exclude_label }}
+    set_fact:
+      worker_node_exclude_list: "{{ exclude_node_list | json_query('results[].resources[].metadata.name') }}"
+  
+  - name: Exclude labeled nodes
+    set_fact:
+        worker_node_list: "{{ worker_node_list | difference(worker_node_exclude_list) }}"
+    when: workload_args.exclude_labels is defined and workload_args.exclude_labels | length > 0
+    #
+    # Compute node and pod limits using CR params while taking into account 
+    # of the actual number of nodes available in the system
+    #
+  - name: init pod and node low/hi idx
+    set_fact:
+        pod_low_idx: "{{ workload_args.density_range[0] | default('1')|int - 1 }}"
+        pod_hi_idx: "{{ workload_args.density_range[1] | default('1')|int - 1 }}"
+        node_low_idx: "{{ workload_args.node_range[0] | default('1')|int - 1 }}"
+        node_hi_idx: "{{ workload_args.node_range[1]  | default('1')|int - 1 }}"
+    #
+    # Next sanity check and massage the indices if necessary.
+    # We shall complete gracefully and not iterate wildly.
+    #
+  - name: Adjust node_hi_idx if cluster has less nodes
+    set_fact:
+        node_hi_idx: "{{ worker_node_list|length| default('0')|int -1 }}"
+    when: "node_hi_idx|int >= worker_node_list|length| default('0')|int "
+
```

Comment:
Maybe we should consider hanging if  # worker nodes < pod_hi_idx

Reply:
"node" goes with "node_range". "node_hi_idx" does not associate with pod/density_range.

Reply:
I actually was suggesting to change the logic in setup.yml to hang if the number of nodes is lower than the higher number of pods to deploy (I think density_range[1]).

Reply:
Robustness is very desirable if not required. This is a universal goal we should continue to try.  Regarding about your statement "number of nodes is lower than the number of pods", this is irrelevant because there is no restriction in that regard. You can run node_range:[2,2], density_range[1,100]. In this case we have 100 pairs per nodes for total of 200 pairs.